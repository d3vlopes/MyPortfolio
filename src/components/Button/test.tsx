import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { Github } from '@styled-icons/boxicons-logos'

import Button from '.'

const { getByRole, getByTestId } = screen

describe('<Button />', () => {
  it('should render by default', () => {
    renderWithTheme(
      <Button variant="primary" size="medium">
        Button
      </Button>,
    )

    const button = getByRole('button', { name: /button/i })

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      'background-color': '#6D59A8',
      'font-size': '2.2rem',
      width: '21.5rem',
      height: '6.2rem',
    })
  })

  it('should render the variant secondary ', () => {
    renderWithTheme(<Button variant="secondary">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      'background-color': '#83A848',
    })
  })

  it('should render the variant outline ', () => {
    renderWithTheme(<Button variant="outline">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      border: '1px solid #DEE7EA',
      'background-color': 'transparent',
    })
  })

  it('should render expand version', () => {
    renderWithTheme(<Button expand>Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      width: '31.5rem',
      height: '6.2rem',
    })
  })

  it('should render small size', () => {
    renderWithTheme(<Button size="small">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      'font-size': '1.8rem',
      width: '17.5rem',
      height: '5rem',
    })
  })

  it('should render large size', () => {
    renderWithTheme(<Button size="large">Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyle({
      'font-size': '2.6rem',
      width: '28.8rem',
      height: '7.6rem',
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<Github data-testid="icon" />}>Button</Button>,
    )

    const button = getByRole('button', { name: /button/i })
    const icon = getByTestId('icon')

    expect(button).toBeInTheDocument()
    expect(icon).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Button
      </Button>,
    )

    expect(screen.getByRole('link', { name: /button/i })).toHaveAttribute(
      'href',
      '/link',
    )
  })

  it('should up while hover', async () => {
    renderWithTheme(<Button hoverUp>Button</Button>)

    const button = getByRole('button', { name: /button/i })

    expect(button).toHaveStyleRule('transform', 'translateY(-0.5rem)', {
      modifier: ':hover',
    })
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(<Button>Button</Button>)

    expect(container.firstChild).toMatchSnapshot()
  })
})
