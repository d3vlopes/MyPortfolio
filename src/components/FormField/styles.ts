import styled, { css, DefaultTheme } from 'styled-components'

type WrapperProps = {
  error?: boolean
  hasFocus?: boolean
}

export const FieldWrapper = styled.div<WrapperProps>`
  ${({ theme, error, hasFocus }) => css`
    position: relative;
    background-color: ${theme.colors.bg.secondary.main};
    border-radius: ${theme.border.radius.small};
    margin-bottom: ${error ? theme.spacings.xxsmall : theme.spacings.small};

    input,
    textarea {
      padding: ${theme.spacings.medium} ${theme.spacings.small};
      background-color: transparent;
      border: 0;
      outline: none;
      width: 100%;
      font-size: ${theme.font.sizes.small};
      font-family: ${theme.font.body};
      line-height: 150%;
      color: ${theme.colors.text.light};

      &:focus ~ label,
      &:valid ~ label,
      ${TextArea}:focus ~ label,
      ${TextArea}:valid ~ label {
        top: ${hasFocus && '-1.8rem'};
        transition: all ${theme.transition.ease.default};
        color: ${theme.colors.text.light};
        font-size: ${theme.font.sizes.xxsmall};
      }

      &:valid {
        border: 1px solid ${theme.colors.complementary.main};
      }
    }
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    &:-webkit-autofill {
      -webkit-text-fill-color: ${theme.colors.text.light};
      border: 1px solid ${theme.colors.complementary.main};
      -webkit-box-shadow: 0 0 0px 1000px ${theme.colors.bg.secondary.main} inset;
    }
  `}
`

export const TextArea = styled.textarea`
  width: 100%;
  min-height: 25rem;
  resize: none;
`

export const Label = styled.label`
  ${({ theme }) => css`
    position: absolute;
    top: 0.1rem;
    left: 0;
    padding: ${theme.spacings.medium} 0 ${theme.spacings.medium}
      ${theme.spacings.small};
    font-weight: ${theme.font.weight.medium};
    font-size: ${theme.font.sizes.small};
    line-height: 100%;
    letter-spacing: 0.005rem;
    color: ${theme.colors.grey};
    pointer-events: none;
    transition: all ${theme.transition.ease.default};
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xxxsmall};
  `}
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${FieldWrapper} input,
    ${FieldWrapper} textarea {
      border-color: ${theme.colors.red};
    }
  `,
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error }) => css`
    width: 100%;

    ${error && wrapperModifiers.error(theme)}
  `}
`
