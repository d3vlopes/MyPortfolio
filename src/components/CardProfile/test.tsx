import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import CardProfile, { CardProfileProps } from '.'

const props: CardProfileProps = {
  image: '/img/me.jpg',
  name: 'Leandro Lopes',
  age: 27,
  city: 'Santa Cruz do Sul - RS',
  liking: 'Programação, design, música, séries e filmes',
  dream:
    'Trabalhar com o que amo, utilizando meu conhecimento para criar aplicações que possam ajudar as pessoas.',
}

const { getByRole, getByText } = screen

describe('<CardProfile />', () => {
  it('should render CardProfile', () => {
    const { container } = renderWithTheme(<CardProfile {...props} />)

    const image = getByRole('img', { name: 'Leandro Lopes' })
    const name = getByText('Leandro Lopes')
    const age = getByText('27 anos')
    const city = getByText('Santa Cruz do Sul - RS')
    const liking = getByText('Programação, design, música, séries e filmes')
    const dream = getByText(
      'Trabalhar com o que amo, utilizando meu conhecimento para criar aplicações que possam ajudar as pessoas.',
    )

    expect(image).toHaveAttribute('src', '/img/me.jpg')
    expect(name).toBeInTheDocument()
    expect(age).toBeInTheDocument()
    expect(city).toBeInTheDocument()
    expect(liking).toBeInTheDocument()
    expect(dream).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
