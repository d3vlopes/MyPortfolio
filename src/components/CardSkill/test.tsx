import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { Sword } from '@styled-icons/remix-fill/Sword'

import CardSkill from '.'
import mock from './mock'

const { getByTestId, getByRole, getByText } = screen

describe('<CardSkill />', () => {
  it('should render CardSkill', () => {
    renderWithTheme(
      <CardSkill
        icon={<Sword data-testid="icon" />}
        title={mock[0].title}
        description={mock[0].description}
      />,
    )

    const icon = getByTestId('icon')
    const title = getByRole('heading', { name: /desafios/i })
    const description = getByText(
      /Estou sempre em busca de novos desafios que me tirem do meu porto seguro./i,
    )

    expect(icon).toBeInTheDocument()
    expect(title).toBeInTheDocument()
    expect(description).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(
      <CardSkill
        icon={<Sword data-testid="icon" />}
        title={mock[0].title}
        description={mock[0].description}
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
