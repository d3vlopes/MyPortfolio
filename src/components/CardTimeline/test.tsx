import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import CardTimeline, { CardTimelineProps } from '.'

const props: CardTimelineProps = {
  title: 'CardTimeline',
  content: '<p>Loren ipsum dolor</p>',
}

const { getByRole, getByText } = screen

describe('<CardTimeline />', () => {
  it('should render CardTimeline', () => {
    const { container } = renderWithTheme(<CardTimeline {...props} />)

    const title = getByRole('heading', { name: 'CardTimeline' })
    const content = getByText('Loren ipsum dolor')
    const p = container.querySelector('p')

    expect(title).toBeInTheDocument()
    expect(content).toBeInTheDocument()
    expect(p?.tagName.toLowerCase()).toBe('p')
  })

  it('should match snapshot ', () => {
    const { container } = renderWithTheme(<CardTimeline {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
