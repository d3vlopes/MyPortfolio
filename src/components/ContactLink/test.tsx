import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { Props } from 'react-inlinesvg'

import ContactLink, { ContactLinkProps } from '.'

const { getByLabelText } = screen

const mock: ContactLinkProps = {
  icon: '/img/icon.svg',
  name: 'Github',
  link: 'www.github.com/user/name',
}

jest.mock('react-inlinesvg', () => {
  function ReactInlineSVG({ name }: Props) {
    return <svg aria-label={name} />
  }

  return ReactInlineSVG
})

describe('<ContactLinks />', () => {
  it('should render ContactLinks', () => {
    const { container } = renderWithTheme(<ContactLink {...mock} />)

    const icon = getByLabelText(mock.name)
    const a = container.querySelector('a')

    expect(icon).toBeInTheDocument()
    expect(a).toHaveAttribute('href', mock.link)
  })
})
