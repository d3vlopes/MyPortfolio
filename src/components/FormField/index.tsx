import {
  Dispatch,
  InputHTMLAttributes,
  useState,
  SetStateAction,
  TextareaHTMLAttributes,
} from 'react'

import * as S from './styles'

export type FormFieldProps = {
  type?: 'input' | 'textarea'
  label: string
  initialValue?: string
  error?: string

  onFieldChange?: (value: string) => void
} & InputHTMLAttributes<HTMLInputElement> &
  TextareaHTMLAttributes<HTMLTextAreaElement>

const FormField = ({
  type = 'input',
  label,
  name,
  initialValue = '',
  error,
  onFieldChange,
  ...props
}: FormFieldProps) => {
  const [inputValue, setInputValue] = useState(initialValue)
  const [textAreaValue, setTextAreaValue] = useState(initialValue)

  const onChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    setValue: Dispatch<SetStateAction<string>>,
  ) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onFieldChange && onFieldChange(newValue)
  }

  const renderInput = () => (
    <S.Input
      type="text"
      onChange={(event) => onChange(event, setInputValue)}
      value={inputValue}
      name={name}
      required
      id={name}
      {...props}
    />
  )

  const renderTextArea = () => (
    <S.TextArea
      value={textAreaValue}
      required
      onChange={(event) => onChange(event, setTextAreaValue)}
      name={name}
      id={name}
      {...props}
    />
  )

  return (
    <S.Wrapper error={!!error}>
      <S.FieldWrapper error={!!error}>
        {type === 'input' ? renderInput() : renderTextArea()}
        {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      </S.FieldWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default FormField
