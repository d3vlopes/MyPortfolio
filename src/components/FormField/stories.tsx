import { Story, Meta } from '@storybook/react'
import FormField, { FormFieldProps } from '.'

export default {
  title: 'Form/FormField',
  component: FormField,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/KIQWZ47GDrrNCIxysbEqOA/My-Portfolio?node-id=813%3A1060',
    },
  },
  args: {
    type: 'input',
    label: 'Email',
    name: 'email',
    initialValue: '',
    disabled: false,
    error: undefined,
    fullWidth: false,
  },
  argTypes: {
    onInput: { action: 'changed' },
  },
} as Meta<FormFieldProps>

export const Default: Story<FormFieldProps> = (args) => {
  return (
    <div style={{ maxWidth: '130rem', margin: '0 auto' }}>
      <FormField {...args} />
    </div>
  )
}

export const withError: Story<FormFieldProps> = (args) => {
  return (
    <div style={{ maxWidth: '130rem', margin: '0 auto' }}>
      <FormField {...args} />
    </div>
  )
}

withError.args = {
  label: 'Email',
  value: 'test.com.br',
  error: 'Ops... something is wrong',
}

export const textarea: Story<FormFieldProps> = (args) => {
  return (
    <div style={{ maxWidth: '130rem', margin: '0 auto' }}>
      <FormField {...args} />
    </div>
  )
}

textarea.args = {
  type: 'textarea',
  label: 'Message',
  value:
    'Lorem ipsum dolor sit lorem ipsum dolor sit lorem ipsum dolor sit lorem ipsum dolor sit',
}
