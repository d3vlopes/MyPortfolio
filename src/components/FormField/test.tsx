import { screen, waitFor } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'
import { ThemeProvider } from 'styled-components'

import userEvent from '@testing-library/user-event'

import theme from 'styles/themes/default'

import FormField from '.'

const { getByLabelText, getByText } = screen

describe('<FormField />', () => {
  it('should render FormField like input', () => {
    const { container } = renderWithTheme(
      <FormField label="Nome" name="name" />,
    )

    const input = container.querySelector('input')

    expect(getByLabelText('Nome')).toBeInTheDocument()
    expect(input?.tagName.toLowerCase()).toBe('input')
    expect(input).toHaveAttribute('id', 'name')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render FormField like textarea', () => {
    const { container } = renderWithTheme(
      <FormField variant="textarea" label="Message" name="message" />,
    )

    const textarea = container.querySelector('textarea')

    expect(getByLabelText('Message')).toBeInTheDocument()
    expect(textarea?.tagName.toLowerCase()).toBe('textarea')
  })

  it('Changes its value when typing', async () => {
    const onInputChange = jest.fn()

    const { container, rerender } = renderWithTheme(
      <FormField
        onFieldChange={onInputChange}
        label="FormField"
        name="FormField"
      />,
    )

    const input = container.querySelector('input')
    const text = 'This is my new text'

    userEvent.type(input!, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onInputChange).toHaveBeenCalledWith(text)

    const onTextAreaChange = jest.fn()

    rerender(
      <ThemeProvider theme={theme}>
        <FormField
          onFieldChange={onTextAreaChange}
          label="FormField"
          name="FormField"
          variant="textarea"
        />
        ,
      </ThemeProvider>,
    )

    const textarea = container.querySelector('textarea')

    userEvent.type(textarea!, text)

    await waitFor(() => {
      expect(textarea).toHaveValue(text)
      expect(onTextAreaChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onTextAreaChange).toHaveBeenCalledWith(text)
  })

  it('Render with error ', () => {
    renderWithTheme(<FormField label="FormField" error="Error message" />)

    expect(getByText('Error message')).toBeInTheDocument()
  })

  it('Is accessible by tab', () => {
    renderWithTheme(
      <FormField label="TextField" name="TextField" id="TextField" />,
    )

    const input = getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })
})
