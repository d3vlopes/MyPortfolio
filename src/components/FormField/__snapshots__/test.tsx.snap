// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> should render FormField like input 1`] = `
.c1 {
  position: relative;
  background-color: #3B3A51;
  border-radius: 0.4rem;
  margin-bottom: 2.4rem;
}

.c1 input,
.c1 textarea {
  padding: 3.2rem 2.4rem;
  background-color: transparent;
  border: 0;
  outline: none;
  width: 100%;
  font-size: 1.8rem;
  font-family: Rubik,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  line-height: 150%;
  color: #FFFFFF;
}

.c1 input:valid,
.c1 textarea:valid {
  border: 1px solid #DEE7EA;
}

.c2:-webkit-autofill {
  -webkit-text-fill-color: #FFFFFF;
  border: 1px solid #DEE7EA;
  -webkit-box-shadow: 0 0 0px 1000px #3B3A51 inset;
}

.c3 {
  position: absolute;
  top: 0.1rem;
  left: 0;
  padding: 3.2rem 0 3.2rem 2.4rem;
  font-weight: 500;
  font-size: 1.8rem;
  line-height: 100%;
  -webkit-letter-spacing: 0.005rem;
  -moz-letter-spacing: 0.005rem;
  -ms-letter-spacing: 0.005rem;
  letter-spacing: 0.005rem;
  color: rgba(255,255,255,0.5);
  pointer-events: none;
  -webkit-transition: all 0.25s ease;
  transition: all 0.25s ease;
}

.c0 {
  width: 100%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2"
      id="name"
      name="name"
      required=""
      type="text"
      value=""
    />
    <label
      class="c3"
      for="name"
    >
      Nome
    </label>
  </div>
</div>
`
